libsolidity/formal/CHC.cpp

```diff
@@ -1164,49 +1223,235 @@ void CHC::checkAssertTarget(ASTNode const* _scope, CHCVerificationTarget const&
-		(void)model;
-		if (result == smtutil::CheckResult::UNSATISFIABLE)
-			m_safeTargets[assertion].insert(_target.type);
+		checkAndReportTarget(assertion, _target, errorId, 6328_error, "Assertion violation happens here");
 	}
 }
 
 void CHC::checkAndReportTarget(
 	ASTNode const* _scope,
 	CHCVerificationTarget const& _target,
 	unsigned _errorId,
+	ErrorId _errorReporterId,
 	string _satMsg,
 	string _unknownMsg
 )
 {
+	if (m_unsafeTargets.count(_scope) && m_unsafeTargets.at(_scope).count(_target.type))
+		return;
+
 	createErrorBlock();
 	connectBlocks(_target.value, error(), _target.constraints && (_target.errorId == _errorId));
-	auto [result, model] = query(error(), _scope->location());
-	// This should be fine but it's a bug in the old compiler
-	(void)model;
+	auto const& [result, model] = query(error(), _scope->location());
 	if (result == smtutil::CheckResult::UNSATISFIABLE)
 		m_safeTargets[_scope].insert(_target.type);
 	else if (result == smtutil::CheckResult::SATISFIABLE)
 	{
 		solAssert(!_satMsg.empty(), "");
 		m_unsafeTargets[_scope].insert(_target.type);
-		m_outerErrorReporter.warning(
-			2529_error,
-			_scope->location(),
-			_satMsg
-		);
+		auto cex = generateCounterexample(model, error().name);
+		if (cex)
+			m_outerErrorReporter.warning(
+				_errorReporterId,
+				_scope->location(),
+				_satMsg,
+				SecondarySourceLocation().append(" for:\n" + *cex, SourceLocation{})
+			);
+		else
+			m_outerErrorReporter.warning(
+				_errorReporterId,
+				_scope->location(),
+				_satMsg + "."
+			);
 	}
 	else if (!_unknownMsg.empty())
 		m_outerErrorReporter.warning(
-			1147_error,
+			_errorReporterId,
 			_scope->location(),
 			_unknownMsg
 		);
 }
 
+/**
+The counterexample DAG has the following properties:
+1) The root node represents the reachable error predicate.
+2) The root node has 1 or 2 children:
+	- One of them is the summary of the function that was called and led to that node.
+	If this is the only child, this function must be the constructor.
+	- If it has 2 children, the function is not the constructor and the other child is the interface node,
+	that is, it represents the state of the contract before the function described above was called.
+3) Interface nodes also have property 2.
+
+The following algorithm starts collecting function summaries at the root node and repeats
+for each interface node seen.
+Each function summary collected represents a transaction, and the final order is reversed.
+
+The first function summary seen contains the values for the state, input and output variables at the
+error point.
+*/
+optional<string> CHC::generateCounterexample(CHCSolverInterface::CexGraph const& _graph, string const& _root)
+{
+	optional<unsigned> rootId;
+	for (auto const& [id, node]: _graph.nodes)
+		if (node.first == _root)
+		{
+			rootId = id;
+			break;
+		}
+	if (!rootId)
+		return {};
+
+	vector<string> path;
+	string localState;
+
+	unsigned node = *rootId;
+	optional<string> lastTxSeen;
```

Comment:
That's basically just a node ``bool lastTx = true;`` with ``if (lastTx) { lastTx = false; ...`` below, right? Maybe with a small comment stating that the first node represents the last transaction...

Reply:
Yea, I think I wanted to keep the actual name or id for something else, but not really needed the way it's written right now.